mirgecom/diffusion.py

```diff
@@ -0,0 +1,111 @@
+r""":mod:`mirgecom.diffusion` computes the diffusion operator.
+
+.. autofunction:: diffusion_operator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import math
+import numpy as np
+import numpy.linalg as la  # noqa
+from pytools.obj_array import make_obj_array, obj_array_vectorize
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from meshmode.dof_array import thaw
+from grudge.symbolic.primitives import TracePair
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+
+
+def _q_flux(discr, alpha, u_tpair):
+    normal = thaw(u_tpair.int.array_context, discr.normal(u_tpair.dd))
+    flux_weak = make_obj_array([math.sqrt(alpha)*u_tpair.avg])*normal
```

Comment:
```suggestion
    flux_weak = scalar([math.sqrt(alpha)*u_tpair.avg])*normal
```
looks less weird to me. How do you feel about that? (`scalar` lives [here](https://github.com/illinois-ceesd/mirgecom/blob/master/mirgecom/euler.py#L168-L170), but it might be better to dump it in some shared utility module)

Reply:
Yep, I can do that.

I was wondering though: now that `DOFArray` is no longer an object array, would it make sense to define multiplication (and other operations) between scalar `DOFArray`s and arrays of `DOFArray`s? Maybe something like this:

```diff
     def _bop(self, f, op1, op2):
         """Broadcasting logic for a generic binary operator."""
+        if isinstance(op1, np.ndarray):
+            return obj_array_vectorize(lambda op: self._bop(f, op, op2), op1)
+        if isinstance(op2, np.ndarray):
+            return obj_array_vectorize(lambda op: self._bop(f, op1, op), op2)
         ...
```
?

It seems to work, at least for this case. I'm currently running through all the meshmode/grudge/mirgecom tests to see if it breaks anything.

Reply:
Hmm. Runs into problems with expressions that use regular non-object numpy arrays, like `np.eye(dim)*scalar(p)`.